src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -3090,20 +3087,36 @@ var projectUniqueNamesForBuildIntToUpdate
                     .Distinct()
                     .ToList();
 
-                var successfulFrameworks = allFrameworks
-                    .Except(unsuccessfulFrameworks)
-                    .ToList();
+                var originalFrameworks = updatedPackageSpec
+                    .TargetFrameworks
+                    .ToDictionary(x => x.FrameworkName, x => x.TargetAlias);
+
+                List<(NuGetProjectAction, BuildIntegratedInstallationContext)> projectActionsAndInstallationContexts = new(nuGetProjectActions.Length);
+
+                foreach (var action in nuGetProjectActions)
+                {
+                    PackageSpec referencePackageSpec = null;
+                    referencePackageSpec = action.NuGetProjectActionType switch
+                    {
+                        NuGetProjectActionType.Install or NuGetProjectActionType.Update => restoreResult.Result.LockFile.PackageSpec,
+                        NuGetProjectActionType.Uninstall => originalPackageSpec,
+                        _ => throw new InvalidOperationException("Unknown NuGetProjectActionType"),
+                    };
+                    BuildIntegratedInstallationContext installationContext = CreateInstallationContextForPackageId(action.PackageIdentity.Id, referencePackageSpec, unsuccessfulFrameworks, originalFrameworks);
+                    projectActionsAndInstallationContexts.Add((action, installationContext));
+                }
 
-                var firstAction = nuGetProjectActionsList[0];
+                var firstInstallationAndProjectContext = projectActionsAndInstallationContexts[0];
+                var firstAction = firstInstallationAndProjectContext.Item1;
+                var successfulFrameworks = firstInstallationAndProjectContext.Item2.SuccessfulFrameworks.AsList();
```

Comment:
If there are multiple items in the collection `projectActionsAndInstallationContexts`, I am curious to learn why we only access first item in the collection.

Reply:
> Do you think there is any technical debt in this code path that requires a closer look at the API? If yes, feel free to create a new technical debt issue.

Maybe?

1 action meaning more packages has permeated throughout the APIs and I'm not really sure if that's necessarily the best design, but it's also not a frequent code path, but the public APIs just make things nearly impossible to change.
I'm gonna look into the whole tech debt thing, and I'll consider this then.